### DISCUSSION TITLE
Azure Function Deployment

### DISCUSSION THREAD

**Discussion by pablo-padua** [developmentseed/titiler#1026] [Q&A] [answered]
Hello,

First of all, thanks for the azure functions deployment option and README file, that consistently helps new developers to learn how to use lambda/functions/cloud deployments... The current README file is a bit outdated though and I have created a branch with updated commands but have no permission to push, so here's my 2 cents regarding the updated commands:

```bash
$ git clone https://github.com/developmentseed/titiler.git
$ cd titiler/deployment/azure

$ az login
$ az group create --name AzureFunctionsTiTiler-rg --location eastus
$ az storage account create --name {your-new-storage-name} --sku Standard_LRS -g AzureFunctionsTiTiler-rg
$ az functionapp create --consumption-plan-location eastus --runtime python --runtime-version 3.9 --functions-version 4 --name {your-new-function-name} --os-type linux -g AzureFunctionsTiTiler-rg -s {your-new-storage-name}
$ func azure functionapp publish titiler --python
```

I changed python runtime version from 3.8 to 3.9 (azure function no longer supports 3.8) and functions version was updated from 3 to 4 (you wont be able to create a new app function on version 3). Here's some testing I made that backs what I am saying:
![image](https://github.com/user-attachments/assets/ab6d121f-3b14-4c78-8e69-2d394a657d1c)
![image](https://github.com/user-attachments/assets/b4da7ff7-889e-4f33-ae5e-246ef2baa820)
![image](https://github.com/user-attachments/assets/c0e2fd4b-996b-46ee-a184-df465f7b5909)
I have added variables instead of fixed values, such as for container name and app name, given that it must be a unique name for each new instance.

Now, to my actual question - I am getting `404` on `/cog/tiles` on my newly deployed Titiler on Azure Functions, but `/cog/preview` works fine. Has anyone experienced this problem yet ?

Not found on /cog/tiles:
![image](https://github.com/user-attachments/assets/233e8b62-52b6-4f36-a6e6-0b2a70572fcc)
But works on /cog/preview
![image](https://github.com/user-attachments/assets/de5041bd-c271-4665-9cfb-6147a62d61f9)

**Comment by hrodmn** [comment]
Hi @pablo-padua! Thanks for working on the Azure function docs. If you haven't already, please create a fork of the repo in your Github Account and open a PR from your fork to this repo. We would really appreciate it!

I think you are getting a 404 with your other request because it is getting the wrong set of path parameters in the URL.

To request a specific `xyz` tile you will need to provide the path parameter for `tileMatrixSetId` (e.g. `WebMercatorQuad`) before the `/{z}/{x}/{y}` path parameters. The whole request will look like `/cog/tiles/{tileMatrixSetId}/16/24580/36818?...`.

I hope that helps!

**Reply by pablo-padua** [reply]
Hi @hrodmn ,
Thanks for the reply. As for the PR, I have opened it [here](https://github.com/developmentseed/titiler/pull/1028). It's a pleasure to contribute, even if it's just for a minor fix.
As for the path parameter for `tileMatrixSetId`, has that recently changed? Because I have a dockerized titiler running on a personal VPS and the same request as above works fine for me

**Reply by hrodmn** [reply, marked as answer]
Yes this was a breaking change introduced in the most recent version ([0.19.0](https://github.com/developmentseed/titiler/blob/main/CHANGES.md#0190-2024-11-07)). If you want to maintain access to tiles via `/tiles/{z}/{y}/{x}` you can pin to version <0.19.

**Reply by pablo-padua** [reply]
Hi @hrodmn!

I've updated my application to use the new URL and it works fine, thank you very much for your help.
I have another question regarding titiler itself, when I call the `/tiles/{z}/{y}/{x}` endpoint, does it call the COG url for every single coordinate request or does it cache and reuse ? I'm asking because I'm using Azure CDN and price skyrocketed with titiler, so I might need a workaround. If it does fetch the COG url per coordinate request, is there any way to cache it and if not, do you recommend using any other service to reduce costs ?

**Reply by hrodmn** [reply]
Hi @pablo-padua - yes, titiler will read directly from the COG for each request to the `/tiles` endpoint but that is a feature of dynamic tiling applications. If you are reading from COGs the read operation should be very efficient, though. If you post more details about the files that you are trying to use we might be able to make some suggestions for improving read efficiency.

You can set up a titiler application with a cache to speed up responses to identical requests ([example](https://developmentseed.org/titiler/examples/code/tiler_with_cache/), but that will not change the function execution count.

I don't have much experience with Azure Functions + CDN deployments so I don't know which factors are most likely to be driving the increase in cost. If it is function execution counts there is probably not much to do to improve costs in this context. if it is function execution time, there *might* be some titiler-side improvements you could make.
